src/v/kafka/server/protocol.cc

```diff
@@ -166,15 +166,13 @@ ss::future<> protocol::apply(net::server::resources rs) {
         config::shard_local_cfg().enable_sasl());
     const auto authn_method = get_authn_method(*rs.conn);
 
-    /*
-     * if sasl authentication is not enabled then initialize the sasl state to
-     * complete. this will cause auth to be skipped during request processing.
-     */
-    security::sasl_server sasl(
-      authn_method == config::broker_authn_method::sasl
-        ? security::sasl_server::sasl_state::initial
-        : security::sasl_server::sasl_state::complete);
+    // Only initialise sasl state if sasl is enabled
+    auto sasl = authn_method == config::broker_authn_method::sasl
```

Comment:
Really like this change, nullopt is much more expressive than claiming that sasl is complete when it's disabled.